### ISSUE TITLE
DDP deadlocks on iteration 2: the gradient sync point releases a device before its peers arrive

### ISSUE THREAD

**Issue by kafonek** [kafonek/burn#1]
`DistributedSyncServer` releases a device from the gradient sync point before its peers arrive, which deadlocks DDP on the second iteration.

### Measurement

One NVIDIA H100 80GB HBM3 per rank, 2 ranks in a single process on one node, driver reporting `cudaDriverVersion 13000`, NCCL `2.23.4+cuda12.8`. `burn` `0.22.0-pre.2`, `cubecl` `0.11.0-pre.2`, CUDA backend, features `train`, `vision`, `metrics`, `ndarray`, `cuda`; `tui` off, fusion default, autotune default. Linux container, `NCCL_NET_PLUGIN=none` and `NCCL_TUNER_PLUGIN=none` (no fabric plugin in the container).

Harness: a small Rust binary wrapping burn's public API. It builds `ExecutionStrategy::ddp(devices, DistributedConfig { all_reduce_op: ReduceOperation::Mean })`, passes it to `SupervisedTraining::with_training_strategy`, and trains a CIFAR-10 convolutional autoencoder (5 strided convs down to 4x4, a linear bottleneck, 3 transposed convs back up, MSE reconstruction loss, Adam). Batch 250 per device.

### Expectation

DDP should run to the end of the epoch. Each rank computes gradients locally, the gradients are all-reduced, the optimizer steps, and the loop continues. Nothing in that sequence should depend on the ranks arriving at the sync point at the same moment.

### Observation

Training reaches `Iteration 2` and stops permanently. No panic, no error, no progress; the process sits there indefinitely.

`gdb -p <pid> -batch -ex "thread apply all bt"` on the hung process shows both DDP worker threads parked in the same place:

```
#5  oneshot::receiver::Receiver::recv<...>
#6  burn_backend::backend::distributed::client::DistributedSyncClient::submit_sync_collective
    at burn-backend-0.22.0-pre.2/src/backend/distributed/client.rs:67
#7  burn_backend::backend::distributed::ops::DistributedOps::submit_sync_collective
#8  burn_autodiff::tensor::AutodiffTensor::backward
#9  burn_dispatch::backend::{impl#2}::backward
#10 burn_tensor::tensor::api::autodiff::backward_impl
#11 burn_tensor::tensor::api::base::Tensor::backward<1>
```

The `DistributedSyncServer` thread is alive and idle, blocked in `Receiver::recv` waiting for a *new* `ActionMessage`. It has processed everything it was sent and will never send another callback. Both GPUs sit at 0% utilisation.

NCCL is healthy at this point. With `NCCL_DEBUG=INFO` both ranks log `ncclCommInitRank ... Init COMPLETE`, exchange 24 matched `Send`/`Recv` pairs for the initial model broadcast, and issue matched `AllReduce` operations up to `opCount 1c`. The divergence is visible on the next one: rank 1 issues `AllReduce opCount 1d` and rank 0 never issues its match.

### Effect

Multi-device DDP training is unusable on the CUDA backend. Any run hangs a few iterations in and holds its GPUs indefinitely, since nothing times out.

### Likely cause

`burn-backend-0.22.0-pre.2/src/backend/distributed/server.rs`:

```rust
fn try_launch_sync(&mut self) {
    if self.all_reduce_ops_queue.is_empty() {
        for d in self.syncing_devices.clone() {
            let callback = self.callbacks.remove(&d.id()).unwrap();
            let closure = Box::new(move || B::sync_collective(&d));
            callback.send(closure).expect("Can send callback");
            self.devices_synced += 1;
        }
        self.syncing_devices.clear();
    }

    if self.devices_synced == self.num_devices {
        self.devices_registered = 0;
        self.devices_synced = 0;
        self.param_required_map.clear();
        self.callbacks.clear();
    }
}
```

A device is released as soon as the all-reduce queue is empty, without waiting for its peers, and the round is only reset once the *last* device syncs. So the first device to arrive returns from `backward()`, runs the next forward and backward, and calls `register_sync_params` for the following round while the previous round is still closing. `devices_registered` is incremented there, so it overshoots `num_devices`. Then the late device arrives and the reset zeroes the counter, discarding the early device's registration.

`launch_ops` gates on an equality test:

```rust
fn launch_ops(&mut self) {
    if self.devices_registered == self.num_devices {
```

Once the counter has been knocked off that value, it never matches again. No all-reduce is launched, `all_reduce_ops_queue` never empties, `try_launch_sync` never releases a callback, and every device blocks in `submit_sync_collective` forever.

Anything that makes ranks arrive at different times widens the window. In practice cubecl's inline autotune does it on the first iterations, which is why the hang lands so early and so reliably.

### Upstream status

I did not find an existing issue describing this. The nearest related material is the multi-node DDP tracking issue and the older collective work, neither of which covers the single-process multi-device sync server.

### Upstream target

`tracel-ai/burn`, crate `burn-backend`.

### Reproduce

Build any model with `ExecutionStrategy::ddp` over two or more CUDA devices in one process and run more than two iterations. A batch size that divides the dataset evenly across devices is required to isolate this (an uneven split hangs later for a different reason, filed separately).

### Not verified

- I have not reproduced this on a non-CUDA backend; the collective path is CUDA-only in 0.22, so I expect it is unreachable elsewhere.
- I have not proven the exact interleaving that knocks the counter off, only that the counter is the gate and that a barrier removes the hang.
- I did not test whether gradient accumulation changes the behaviour.

### Fix

kafonek/burn#4
